tests/dns_name_tests.rs

```diff
@@ -399,10 +401,102 @@ fn dns_name_ref_try_from_ascii_test() {
         .chain(IP_ADDRESS_DNS_VALIDITY.iter())
     {
         assert_eq!(
-            webpki::DnsNameRef::try_from_ascii(s).is_ok(),
+            DnsName::<&[u8]>::try_from_punycode(s).is_ok(),
             is_valid,
-            "DnsNameRef::try_from_ascii_str failed for \"{:?}\"",
+            "DnsName::try_from_punycode failed for \"{:?}\"",
             s
         );
     }
 }
+
+const DNS_NAME_LOWERCASE_TEST_CASES: &[(&str, &str)] = &[
+    // (expected_lowercase, input)
+    ("abc", "abc"),
+    ("abc", "Abc"),
+    ("abc", "aBc"),
+    ("abc", "abC"),
+    ("abc1", "abC1"),
+    ("abc.def", "abC.Def"),
+];
+
+#[test]
+fn test_dns_name_ascii_lowercase_chars() {
+    for (expected_lowercase, input) in DNS_NAME_LOWERCASE_TEST_CASES {
+        let dns_name: DnsNameRef = DnsName::try_from_punycode(*input).unwrap();
+        let actual_lowercase = dns_name.punycode_lowercase_bytes();
+
+        assert_eq!(expected_lowercase.len(), actual_lowercase.len());
+        assert!(expected_lowercase
+            .chars()
+            .zip(actual_lowercase)
+            .all(|(a, b)| a == b.into()));
+    }
+}
+
+// XXX: This shouldn't require the "std" feature.
+#[cfg(feature = "std")]
+#[test]
+fn test_dns_name_fmt() {
+    for (expected_lowercase, input) in DNS_NAME_LOWERCASE_TEST_CASES {
+        let dns_name: DnsNameRef = DnsName::try_from_punycode(*input).unwrap();
+
+        // Test `Display` implementation.
+        assert_eq!(*expected_lowercase, format!("{}", dns_name));
+
+        // Test `Debug` implementation.
+        let debug_formatted = format!("{:?}", &dns_name);
+        assert_eq!(
+            String::from("DnsName(\"") + *expected_lowercase + "\")",
+            debug_formatted
+        );
+    }
+}
+
+// XXX: We need more test cases.
+#[test]
+fn test_dns_name_eq_different_len() {
```

Comment:
:sigh: Supporting `Hash` and `PartialEq` for `DnsName` has high cost for the benefit. I'll probably spend more time writing these equality/hash tests than doing all the rest.